man/update_server.8
.TH UPDATE_SERVER 8 "30 Nov 1988" "Project Athena"
\" RCSID: $Header: /afs/athena.mit.edu/astaff/project/moiradev/repository/moira/man/update_server.8,v 1.4 1993-10-21 14:39:09 mar Exp $
.SH NAME
update_server \- daemon that accepts config files from Moira
.SH SYNOPSIS
.B update_server
.SH DESCRIPTION
An
.B update_server
must be started at boot time on each server that is to be updated by
Moira.  It should be run as root.  It takes no arguments, and will
detach from the tty and put itself in background.  It attempts to log
its actions to stdout, so the output should be directed to a file if
it is needed.
.SH OPTIONS
While there are no command line options, a configuration file
.I /etc/athena/moira.conf
may specify a number of options.  This file may contain blank lines,
comments preceeded by hash marks, boolean options, or string options.
A boolean option is set by just putting the name of the option on a
line.  A string option is specified by putting the name of the option,
some whitespace, and then the value.
.IP "user username"
This will set the UID of the running server to the specified user's
UID.  All file accesses and the execution of the instructions will be
with this user's privileges.
.IP "chroot directory"
This will set the root directory to the specified directory for the
running of the server.  It uses the chroot system call, so when used
in conjunction with the
.I user
option should make the server quite secure.
.IP nofork
This boolean specifies that the server should not fork when a new
connection is opened.  This is only useful for debugging, as when the
client connection is closed, the server will exit.
.IP noclobber
This boolean specifies that the server is not allowed to overwrite any
existing files while it is receiving datafiles and instructions.
.IP noexec
This boolean specifies that the server is not allowed to execute any
instructions.  It may receive datafiles and instructions, but not
execute anything.
.IP "auth user.instance@REALM"
This specifies a Kerberos principal that is allowed to connect to the
server.  This option may be specified multiple times with different
principals.  If none are specified, the user
.B sms
in the local realm is assumed.
.SH FILES
/etc/athena/moira.conf \- For configuration variables.
/etc/athena/srvtab \- It must be able to get rcmd Kerberos tickets.
.SH "SEE ALSO"
The Project Athena Technical Plan section on Moira.


debian/changelog
debathena-moira (4.0.0+cvs20090409-0debathena1) unstable; urgency=low

  * New CVS snapshot (Trac: #195)
  * Drop patches that have been incorporated upstream.
  * Add patch to allow building on systems without krb4. (Trac: #198)

 -- Evan Broder <broder@mit.edu>  Sun, 12 Apr 2009 14:03:22 -0400

debathena-moira (4.0.0+cvs20080830-0debathena2) unstable; urgency=low

  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 15:32:35 -0500

debathena-moira (4.0.0+cvs20080830-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Add a debathena-moira-update-server package that installs
    update_server

 -- Evan Broder <broder@mit.edu>  Fri, 01 Aug 2008 10:54:26 -0400

debathena-moira (4.0.0+cvs20080701-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Move Moira header files from /usr/share/include to /usr/include/moira

 -- Evan Broder <broder@mit.edu>  Tue, 01 Jul 2008 20:38:13 -0400

debathena-moira (4.0.0+cvs20080304-0debathena1) unstable; urgency=low

  * New CVS snapshot to fix chsh.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 04 Mar 2008 15:50:31 -0500

debathena-moira (4.0.0+cvs20080301-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Remove mrtest segfault patch from last release (fixed in upstream)

 -- Tim Abbott <tabbott@mit.edu>  Sat,  1 Mar 2008 18:51:40 -0500

debathena-moira (4.0.0+cvs20070817-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Patch an mrtest segfault with too many command line arguments.

 -- Anders Kaseorg <andersk@mit.edu>  Fri, 24 Aug 2007 19:53:49 -0400

debathena-moira (4.0.0+cvs20070129-0debathena3) unstable; urgency=low

  * Move chsh.debathena to chsh.moira and chfn.debathena to chfn.moira.

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 06 Aug 2007 03:17:10 -0400

debathena-moira (4.0.0+cvs20070129-0debathena2) unstable; urgency=low

  * Add missing build dependencies to libmoira-dev.
  * Section: debathena/net.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 17 Jul 2007 06:34:07 -0400

debathena-moira (4.0.0+cvs20070129-0debathena1) unstable; urgency=low

  * Initial release.

 -- Anders Kaseorg <andersk@mit.edu>  Wed, 14 Feb 2007 00:53:24 -0500


update/update_server.c
/* $Id: update_server.c,v 1.26 2006-08-22 17:36:26 zacheiss Exp $
 *
 * Copyright 1988-1998 by the Massachusetts Institute of Technology.
 * For copying and distribution information, please see the file
 * <mit-copyright.h>.
 */

#include <mit-copyright.h>
#include <moira.h>
#include "update_server.h"

#include <sys/stat.h>
#include <sys/utsname.h>
#include <sys/wait.h>

#include <netinet/in.h>
#include <arpa/inet.h>

#include <errno.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <syslog.h>

#include <des.h>
#include "update.h"

RCSID("$Header: /afs/athena.mit.edu/astaff/project/moiradev/repository/moira/update/update_server.c,v 1.26 2006-08-22 17:36:26 zacheiss Exp $");

char *whoami, *hostname;

int have_authorization = 0;
des_cblock session;
int uid = 0;

void child_handler(int signal);
static void syslog_com_err_proc(const char *progname, long code,
				const char *fmt, va_list args);

struct _dt {
  char *str;
  void (*proc)(int, char *);
} dispatch_table[] = {
  { "AUTH_002", auth_002 },
  { "AUTH_003", auth_003 },
  { "XFER_002", xfer_002 },
  { "XFER_003", xfer_003 },
  { "EXEC_002", exec_002 },
  { "quit", quit },
  { NULL, (void (*)(int, char *))abort }
};

int main(int argc, char **argv)
{
  char *str, *p;
  size_t len;
  struct _dt *d;
  struct utsname name;
  int s, conn;
  struct sigaction sa;

  whoami = strrchr(argv[0], '/');
  if (whoami)
    whoami++;
  else
    whoami = argv[0];

  /* interpret arguments here */
  if (argc != 1)
    {
      fprintf(stderr, "Usage:  %s\n", whoami);
      exit(1);
    }

  if (!config_lookup("nofork"))
    {
      if (fork())
	exit(0);
      setsid();
    }

  uname(&name);
  hostname = name.nodename;

  umask(0022);
  mr_init();

  sigemptyset(&sa.sa_mask);
  sa.sa_flags = SA_RESTART;
  sa.sa_handler = child_handler;
  sigaction(SIGCHLD, &sa, NULL);

  /* If the config file contains a line "user username", the
   * daemon will run with that user's UID.
   */
  if ((p = config_lookup("user")))
    {
      struct passwd *pw;
      pw = getpwnam(p);
      if (!pw)
	{
	  com_err(whoami, errno, "Unable to find user %s\n", p);
	  exit(1);
	}
      uid = pw->pw_uid;
    }

  /* If the config file contains a line "port portname", the daemon
   * will listen on the named port rather than SERVICE_NAME ("moira_update")
   */
  if (!(p = config_lookup("port")))
    p = SERVICE_NAME;

  s = mr_listen(p);
  if (s == -1)
    {
      com_err(whoami, errno, "creating listening socket");
      exit(1);
    }

  set_com_err_hook(syslog_com_err_proc);
  openlog(whoami, LOG_PID, LOG_DAEMON);

  /* now loop waiting for connections */
  while (1)
    {
      struct sockaddr_in client;
      long len;
      char *buf;

      conn = mr_accept(s, &client);
      if (conn == -1)
	{
	  com_err(whoami, errno, "accepting on listening socket");
	  exit(1);
	}
      else if (conn == 0)
	continue;

      if (config_lookup("nofork") || (fork() <= 0))
	break;

      close(conn);
    }

  /* If the config file contains a line "chroot /dir/name", the
   * daemon will run chrooted to that directory.
   */
  if ((p = config_lookup("chroot")))
    {
      if (chroot(p) < 0)
	{
	  com_err(whoami, errno, "unable to chroot to %s", p);
	  exit(1);
	}
    }

  com_err(whoami, 0, "got connection");

  while (1)
    {
      char *cp, *str;
      size_t len;
      int code;

      code = recv_string(conn, &str, &len);
      if (code)
	{
	  com_err(whoami, code, "receiving command");
	  close(conn);
	  exit(1);
	}

      cp = strchr(str, ' ');
      if (cp)
	*cp = '\0';
      for (d = dispatch_table; d->str; d++)
	{
	  if (!strcmp(d->str, str))
	    {
	      if (cp)
		*cp = ' ';
	      (d->proc)(conn, str);
	      goto ok;
	    }
	}
      com_err(whoami, 0, "unknown request received: %s", str);
      code = send_int(conn, MR_UNKNOWN_PROC);
      if (code)
	com_err(whoami, code, "sending UNKNOWN_PROC");
    ok:
      free(str);
    }
}

int send_ok(int conn)
{
  return send_int(conn, 0);
}

/*
 * quit request:
 *
 * syntax:
 * >>> quit
 * <<< (int)0
 * any arguments are ignored
 *
 * function:
 *	closes connection from MR
 */

void quit(int conn, char *str)
{
  send_ok(conn);
  close(conn);
  com_err(whoami, 0, "Closing connection.");
  exit(0);
}

void fail(int conn, int err, char *msg)
{
  com_err(whoami, err, msg);
  close(conn);
  exit(1);
}

void child_handler(int signal)
{
  int status;

  while (waitpid(-1, &status, WNOHANG) > 0)
    ;
}

static void syslog_com_err_proc(const char *progname, long code,
				const char *fmt, va_list args)
{
  char buf[BUFSIZ + 1];

  buf[BUFSIZ] = '\0';

  vsnprintf(buf, BUFSIZ, fmt, args);
  syslog(LOG_NOTICE, "%s: %s %s", progname ? progname : "",
	 code ? error_message(code) : "", buf);
}
